### PR TITLE
refactor: using renderer2 + universal compatible refactoring

### PR DESCRIPTION
**Pull request by joejordanbrown**
### Refactor using renderer2 and making universal compatible

As discussed in #4 I've tested it a number of times using SSR and it seems to work fine. It still needs a few tweaks in relation to how to calculate scrollbar value when rendering on the server, atm it defaults to 0, but this may not be the best default due to the scrollbar is shown for a split second on load. It might be possible to calculate the default scrollbar size depending on the user agent. Not looked into this yet.

1. If you can create a branch for me for this refactoring I can then edit this pull request to work with that branch instead of master, then we can work from that branch until this is released.

3. I've also updated the packages to the latest versions and made sure that everything compiles correctly. It might be a good idea to review how the build system currently works and follow the [Angular Package Format v4.0](https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/edit#heading=h.k0mh3o8u5hx). Could use a modified version of https://github.com/filipesilva/angular-quickstart-lib.

3. If you could test this and also review that I have it in sync with your latest commits. Then we can see where to go from here. I have renamed some of the private names to be consistent throughout and also changed file names to follow a more structured format. I think the angular2 part should possibly be changed from `angular2-drag-scroll` to `angular-drag-scroll` to follow the angular version naming guidelines.

4. Tests need to be updated to use the new changed structure.

Demo link: https://uixd.co.uk/open-source-software/angular-cli-universal/examples/drag-scroll

### CODE REVIEW

**Comment by bfwg** [comment]
@joejordanbrown 

Nice work, I'll review it later tonight. Meanwhile, do you mind to fix the unit tests? 
`ng test` is the command I believe.

Thanks!

**Comment by joejordanbrown** [comment]
@bfwg 

Sounds good.

Yes, I planned to fix them tonight but I got stuck completing another task and now it's 02:26. I will hopefully get around to that tomorrow day time. I need to look at why test 2 fails, it seems that something has changed with the `overflow` because it's expected `auto` to be `hidden`. I'll have to compare with your new changes to see if I made a mistake with that.

**Comment by bfwg** [comment]
@joejordanbrown 

Take your time.

The code is looking good so far. I noticed a small thing: when you remove `drag-scroll-snap` attribute from the children element in the demo app, the "move right" button is disabled. 

Should we add the functionality to move the carousel in a non-snap way when the attribute is off?

**Comment by joejordanbrown** [comment]
@bfwg 

I've just added the feature to move left and right the carousel when in no-snap mode. 

I've also spotted that the method `setNavStatus()` is being called a lot which could affect performance. I've added a console.log to show how many times it's being called, try to use the buttons left and right or dragging the UI. You will see it being called 100s of times on a drag. So that should be refactored to only be called on scroll completed with mouse event or transition.

I'm going to look into the tests shortly to fix the problem with the `overflow` because it's expected `auto` to be `hidden`.

**Comment by bfwg** [comment]
@joejordanbrown 

Nice, the unit tests are passing! I don't know how to make changes directly to this PR branch. However, I have cloned the PR branch and made the change to resolve the `setNavStatus()` performance issue. Can you put the changes in https://github.com/bfwg/angular2-drag-scroll/commit/2c0408b19e9e3a2613fb1591180a2a6ebf9967ae and give it a test?

File: src/drag-scroll.ts [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    const childrenArray = this.snaps['_results'];
+    const contentElement = this._contentRef.nativeElement;
+    if (!this.scrollReachesRightEnd && this.snaps.length !== 0 && childrenArray[this.currIndex + 1]) {
+      this.currIndex++;
+      clearTimeout(this.scrollToTimer);
+      this.scrollTo(contentElement, this.toChildrenLocation(), 500);
+    } else if (this.snaps.length === 0) {
+      this.currIndex++;
+      clearTimeout(this.scrollToTimer);
+      this.scrollTo(contentElement, this._contentRef.nativeElement.scrollLeft + 200, 500);
```
Comment (bfwg) [inline src/drag-scroll.ts]: Great job on making the navigation buttons working with no snaps. I wonder what if it is possible to move to the position of the next item instead of a fixed position.

Comment (joejordanbrown) [reply src/drag-scroll.ts]: Yes, I was going to do that after you confirmed that it was fine. I'll see what the best approach is for getting the children without using `nativeElement`. I think I have a way to do that, will need to test.

**Comment by joejordanbrown** [comment]
@bfwg 

I've pulled your changes into this branch. 

I will have to keep looking into the best way to get children elements without a directive for non-snap.

**Comment by joejordanbrown** [comment]
@bfwg 

The best approach I can see for this is adding another directive so you can follow the Angular way using renderer. I'll add one called `drag-scroll-item.ts`, with the selector `[dragScrollItem]`. It will then first check for ones with `[dragScrollSnap]` -> `[dragScrollItem]` -> []. Scrolling will work for all but only the first 2 will scroll to the previous or next element exactly.

There is another way to use `[dragScrollSnap]="false"` and add a `@Input` to that and then do extra checks where it checks if it has that directive and if it's false or true. You still have access to the children. Let me know which you preferer. I'll try both now to see if there are any problems that arise.

**Comment by joejordanbrown** [comment]
I have added the 2nd way by using `[dragScrollSnap] || [dragScrollSnap]="true"` and `[dragScrollSnap]="false"`

I've also added server-side scrollbar size detection using the browser user agent. You won't see the scrollbar anymore on the initial server-side render. Looks a lot better.

**Comment by bfwg** [comment]
@joejordanbrown 

Sweet, is `dragScrollSnap` a mandatory field? Since without it the navigation buttons won't work properly. I'd suggest if we are going to make the attribute on children selectors a mandatory field, we should rename is to something more generic like `dragScrollItem` and make an optional attribute `dragScrollItemSnap` if the user wants drag n snap functionality. 

I like your first idea better
> The best approach I can see for this is adding another directive so you can follow the Angular way using renderer. I'll add one called drag-scroll-item.ts, with the selector [dragScrollItem]. It will then first check for ones with [dragScrollSnap] -> [dragScrollItem] -> []. Scrolling will work for all but only the first 2 will scroll to the previous or next element exactly.

If you feel that the current implementation is better, please let me know too!

Btw, I'm getting console error while running your latest branch.
image

**Comment by joejordanbrown** [comment]
@bfwg 

No `dragScrollSnap` is not required. The way I've made it you can either have no directive or `dragScrollSnap`. By default `dragScrollSnap` option is `true`, but can be disabled using `dragScrollSnap="false"`. With it being false you won't get the snap action but you will get the exact `scrollTo` when using left and right buttons.


The problem with the first idea it adds more complexity to the component than is needed. I tested both ways and found the 2nd way the best way to go to be clean.


That error is to do with the browser user agent part. I've only tested that on all windows browsers. I was about to test it on OSX and Linux later. What's OS are you using? Can you show me the userAgent object? I have them in the console for testing. It will help me to solve the issue.

**Comment by bfwg** [comment]
@joejordanbrown 

I see, I couldn't run the demo app to test your change and it is hard to visualize just from reading the code. The error is coming from Chrome on my Mac book pro. Please let me know if you need more information.

**Comment by joejordanbrown** [comment]
@bfwg 

`ng serve` worked ok? Is it just while loading you get that error? Look at the console you will see the object with userAgent in? I need to see that object.

I think I know what it was if you try and pull the new changes.

image

**Comment by joejordanbrown** [comment]
@bfwg 

All should be fixed now.

**Comment by bfwg** [comment]
@joejordanbrown  

`ng serve` has no error, here is the userAgent object in Chrome dev-tool. I'm still getting the same error.

**Comment by joejordanbrown** [comment]
@bfwg 

Are you sure you have pulled the latest changes?

I changed `osx` to `mac` on the beloew object.
```
    const scrollbarWidth = {
      mac: {
        'chrome': '15px',
        'safari': '15px',
        'firefox': '15px',
      },
      windows: {
        'chrome': '17px',
        'firefox': '17px',
        'ie': '17px',
        'ms-edge': '12px',
      },
      linux: {
        'chrome': '15px',
        'firefox': '15px',
      }
    };
```

**Comment by bfwg** [comment]
@joejordanbrown 

I got the new branch now, the error is gone 🎉 . However, if the children elements doesn't have the attribute `drag-scroll-snap`, the navigation button will only move to a fixed position. Also, scroll action still snaps when `drag-scrool-snap` is false.

I think the desired functionality should be:
**with `drag-scrool-snap` attribute on children elements:** 
- navigation buttons will move by one item.
- drag snaps
- scroll snaps
 
**without `drag-scrool-snap` attribute or `drag-scrool-snap=false` on children elements:** 
- navigation buttons will move by one item.
- drag should not snap
- scroll should not snap

**Comment by joejordanbrown** [comment]
@bfwg 

Yes, that's the desired action. Without having a directive you can't get touch the `<ng-content>` children's dom. So it's always going to be a fixed scroll in pixels. This could be changed with an option that can override the default `200px`. I've just tested with `drag-scrool-snap="false"` and it doesn't snap. So that's strange if it is on yours.

The current setup works exactly like that apart from without `drag-scrool-snap`, you can see how the `@angular/material` components are built. They follow this approach so the components will be compatible on all platforms. They may add more functionality in the future to Angular version 5, they look like they are changing a lot of how Universal works in that version. Which will improve the Universal Virtual DOM.

This may help with some reasoning, [Amazing Angular2 DOM Tips, Tricks, and Warnings](https://blog.dmbcllc.com/amazing-angular2-dom-tips-tricks-and-warnings/), [Can you satisfy all DOM manipulation needs through Angular 2's abstractions? ](https://www.reddit.com/r/Angular2/comments/5gc53f/can_you_satisfy_all_dom_manipulation_needs/), [Angular 2 plans for Virtual DOM](https://github.com/angular/angular/issues/561). You will be able to find more discussions about this throughout the Angular GitHub issues pages.

**Comment by bfwg** [comment]
@joejordanbrown 

You can reproduce the snap issue by setting `drag-scrool-snap=false` and run the demo app in a mobile device. Upon touch scroll, the carousel snaps.

Since we can't navigate to an item's position without the `drag-scrool-snap` attribute. I would suggest to make `drag-scroll-snap` a mandatory field and the developer should apply the attribute on all the children always. In fact, we should rename the attribute to `drag-scroll-item` to make more sense. Although this is not as intuitive to use from a developer perspective, I feel if that is what it takes to write the carousel in an "Angular" way, we should do it.

**Comment by joejordanbrown** [comment]
@bfwg 

I see I forgot to add the check in the drag event method.

The only reason I made it work without also is for backwards compatibility. I think maybe add a warning if it's not added? Should it really be mandatory if the person just places 1 image or lots of text as the content? Let me know what you think and I'll make the changes tomorrow. Once it's finalised I'll update documentation.

Should we maybe think of other tests also to test snap enabled or not? To prevent future changes breaking such features. Would be good to have.

**Comment by bfwg** [comment]
@joejordanbrown 

Yes, I'd like to follow the guidelines for Component/Directive selectors. 
Something like this:
```html
<drag-scroll>
    <drag-scroll-item drag-scroll-snap="false">
        <img src="image.png">
    </drag-scroll-item>
    <drag-scroll-item>
        <h1>Header<h1>
        <p>
            Content.
        </p>
    </drag-scroll-item>
<drag-scroll>
```
But meanwhile we should keep the attribute selector just for backward compatibility. I'm thinking maybe a `console.warn` to notify the developers that the `@directive` style is deprecated?

**Comment by joejordanbrown** [comment]
@bfwg 

Yes looks good to me, I'll have to rework a few things and make the Directive into a Component so we can use `<drag-scroll-item>` and add the `@Input('drag-scroll-snap')` it's not a lot of work tbh to get that to work.  Yes I'll keep the attr selector for backwards compatibility. The one warning we will have to tell people is that if they ever used `drag-scroll` directly onto a Component then the new update wont work because now `drag-scroll` is a component and you can't have 2 Components on an element. Shouldn't be a big problem since this will be released as 2.* because of the breaking changes.

Yes a `console.warn` should be feasible.

Yes I'll update the `README.md` and `home.component.html` with the relevant new API spec.

Have you had a think about the build system being changed? Following the [Angular Package Format v4.0](https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/edit#heading=h.k0mh3o8u5hx). I've came across this package called [`ng-packagr`](https://github.com/dherges/ng-packagr) which the title says "Compile and package a TypeScript library in Angular Package Format" they have a blog post [here](https://medium.com/spektrakel-blog/angular-libraries-are-fun-fece73cceb05) which is a good read.


I won't be able to make the changes until most likely Monday, I have a few projects to complete over the weekend that are urgent. I'll see if i can find the time tomorrow if possible.

**Comment by bfwg** [comment]
@joejordanbrown 

Sounds good to me. I'll take a look at `ng-packger` tomorrow. I agree, we need to make some changes to how we build this library. If it is ok with you, I would suggest to address the build system changes in a different Pull Request. That way, the scope of this PR is relevant to the topic.

Have a great rest of your weekend.

**Comment by joejordanbrown** [comment]
@bfwg 

Sorry for taking so long to reply, been overloaded with work.

Yes, I agree with that it should be a separate issue.

Could you create a new branch which contains an exact copy of the master branch called "v2 - Refactor", I can then change this pull request from `master` to `v2 - Refactor`, we can then merge this branch into that so it doesn't touch master. Then the new issues can be targeted at that branch and also both of us can work from that branch, I will just make sure I rebase my branch whenever you make changes to it.

I will look into the issues this weekend I should have a little free time to review and add to them. It probably won't be today unless I complete all my other tasks.

**Comment by bfwg** [comment]
@joejordanbrown 

No worries, I have created two more branches for you to rebase:
- `develop` branch (default)
- `v2-refactor`

No rush, I'll be around to do reviews for you in the weekend.

**Comment by joejordanbrown** [comment]
I will make the changes now to get this finalised. The changes that have been added to the develop branch are not really needed to be merged into this branch, are they? I had a quick look over them. If they are extra features we can add them once the main required changes to this have been added. Then we can get this published and work on the new build system issues.

With the naming should it be `ngx-drag-scroll`? We can discuss that on the issue #40

**Comment by joejordanbrown** [comment]
@bfwg 

That's now been made to work as followed

Scroll Snap enabled by default
```html
<drag-scroll>
    <drag-scroll-item>
        <img src="image.png">
    </drag-scroll-item>
    <drag-scroll-item>
        <img src="image.png">
    </drag-scroll-item>
<drag-scroll>

```

Scroll Snap disabled by attribute `dragScrollSnap`, it uses camelCase which follows the guidelines for attributes that take a value. We should also do this with all other attribute options.
```html
<drag-scroll>
    <drag-scroll-item dragScrollSnap="false">
        <img src="image.png">
    </drag-scroll-item>
<drag-scroll>
```

Let me know what you think. If everything is ok you can accept the pull request so it's merged into v2-refactor and then we can work on that branch for the other issues before it is released.

**Comment by bfwg** [comment]
@joejordanbrown 

The code in the demo home component looks good. I really like how the "Angular" way works. The new changes on `develop` branch are mostly bug fixes. Once this is ready, we can merge this into `v2-refactor` branch and then merge it with `develop`.

While I was running `npm start`, I got this error.
```
ERROR in /Users/fan.jin/bfwg/angular2-drag-scroll/demo/app/home/home.component.ts (3,10): Module '"/Users/fan.jin/bfwg/angular2-drag-scroll/node_modules/@angular/material/material"' has no exported member 'MdIconRegistry'.
ERROR in /Users/fan.jin/bfwg/angular2-drag-scroll/demo/app/app.module.ts (6,10): Module '"/Users/fan.jin/bfwg/angular2-drag-scroll/node_modules/@angular/material/material"' has no exported member 'MdButtonModule'.
ERROR in /Users/fan.jin/bfwg/angular2-drag-scroll/demo/app/app.module.ts (6,26): Module '"/Users/fan.jin/bfwg/angular2-drag-scroll/node_modules/@angular/material/material"' has no exported member 'MdSlideToggleModule'.
ERROR in /Users/fan.jin/bfwg/angular2-drag-scroll/demo/app/app.module.ts (6,47): Module '"/Users/fan.jin/bfwg/angular2-drag-scroll/node_modules/@angular/material/material"' has no exported member 'MdToolbarModule'.
ERROR in Error: MdToolbarModule is not an NgModule
```
I think this is breaking the unit test as well.

**Comment by joejordanbrown** [comment]
@bfwg 

Sounds perfect.

Sorry I was in the process of updating all the packages and pushed them but hadn't pushed the Angular Material changes since I had to rush off. I will just test them and push the other changes to fix the deprecated material modules.

**Comment by joejordanbrown** [comment]
@bfwg 

It's now up to date with using the latest packages of Angular and Material which now uses `mat` instead of `md`

**Comment by joejordanbrown** [comment]
@bfwg 

I've fixed the tests and found a slight bug that I'm not sure if it should be fixed to set snap to false if only 1 item or just console.warn. It relates to when you only have 1 `<drag-scroll-item>` if you don't set the `dragScrollSnap` to `false` it will cause an error when it checks if the next item is enabled because it doesn't exist.

**Comment by bfwg** [comment]
@joejordanbrown 
Looking good, I'll do the review later tonight.

File: src/drag-scroll.ts [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+    if (this.currIndex !== 0 && this.children.length !== 0 && this.children['_results'][this.currIndex].enabled === true) {
+      // reach left most
+      this.currIndex--;
+      clearTimeout(this.scrollToTimer);
+      this.scrollTo(contentElement, this.toChildrenLocation(), 500);
+    } else if (this.currIndex !== 0 && this.children.length !== 0 && this.children['_results'][this.currIndex].enabled === false) {
+      this.currIndex--;
+      clearTimeout(this.scrollToTimer);
+      this.scrollTo(contentElement, this.toChildrenLocation(), 500);
+      console.log('dragScrollSnap="false"');
```
Comment (bfwg) [inline src/drag-scroll.ts]: Could we remove all the debug logs?

**Comment by bfwg** [comment]
@joejordanbrown 

This is looking very good, great work! I left some tiny notes in my review. I'll merge this once those are fixed.

**Comment by joejordanbrown** [comment]
@bfwg 

Completed requested changes.

**Comment by bfwg** [comment]
@joejordanbrown  Nice job 🎉 
The change is now in `v2-refactor` branch.

**Comment by joejordanbrown** [comment]
@bfwg 

Great news!

Let me know where you want to go from here. Just tag me in the other issues and we can discuss the build system etc.
